### ISSUE TITLE
mixed models toys for lme4 model inclusion

### ISSUE THREAD

**Issue by bachlaw** [topepo/caret#257]
It would be great if caret could handle mixed models, given the importance at times of modeling observation groups and/or to determine if partial pooling would benefit a model. 

To help this happen, I've created two toy data sets, both based on the high-scoring 1998 season in baseball.  Both are designed to be used with the widely-accepted "lme4" package, although they should work with any mixed-modeling package. These data sets use "batter" and "pitcher" as random effects, although the same approach could be used with observation groups, cities, schools, etc.  So, I feel these are representative and useful.

The first data set is designed for a linear mixed model with crossed and unbalanced random effects.  It contains two acceptable fixed effects: one categorical (stadium) and one continuous (game time temperature, a/k/a "temp_log"); and two categorical random effects (batter and pitcher).  The output, "lwts", is continuous, and refers to the average linear weight of each batting event (a single, a home run, an out, etc.)

The recommended model specification, under the package lme4, is:
lmer(lwts ~ stadium + temp_log + (1|batter) + (1|pitcher))

The second data set is similar except it has a binomial output and thus is a generalized linear mixed model.  To provide the binomial output, this data set focuses on home runs, with "1" being a home run and "0" being a plate appearance with some other outcome.  The recommended specification is:

glmer(HR ~ stadium + temp_log + (1|batter) + (1|pitcher), family=binomial(link='probit'), nAGQ=0).

The standard logit link works fine as well, but in my opinion, the probit link tends to be a better choice for generalized mixed models.  I would suggest supporting both.  

Both toys are in a Zip file entitled "mixed.model.toys.zip" and can be found here: http://1drv.ms/1PdmxDN 

Both data sets are also large, ~ 180k observations.  For ease of modeling, a random subset of half or even less of each file would probably be sufficient.

For those interested, the data sets and models are an outgrowth of our mixed model work at Baseball Prospectus on catcher framing (http://www.baseballprospectus.com/article.php?articleid=25514) and pitcher modeling (http://www.baseballprospectus.com/article.php?articleid=26195).

**Comment by topepo** [comment]
I'm all for this but there are some roadblocks:

1. the resampling should happen at the level of the independent experimental unit(s) such as batter and pitcher in your example. This probably isn't too much of an issue but there is no built-in procedure for doing this (as there are in the usual and time-series cases)
2. there is probably literature about how to resample these models and that should be examined. For example, there may be the need to sample the residuals (or something similar) underneath the experimental units. I haven't thought about this in a long time so I don't really know either way.
3. the biggest technical impediment is the nature of mixed model formulas. They are much richer in functionality than the usual formulas (e.g. `y~x+z`). Using the formula method with `train` right now would probably fail with those formulas since their structure is tied to how the mixed models are specified and not just to how the predictors enter the model. There is a [similar issue with some other machine learning models](http://topepo.github.io/caret/custom_models.html#Illustration3) and my (crappy) approach right now is to manually work it out in the custom methods. You could most likely do this with mixed models, but that means that there would not be an _bona fide_ `train` method for mixed models. 

Of these, the last is the biggest issue. I don't know of a way to have non-standard formulas that would work for mixed models (and the other example that I linked to above). We could make another function that would do exactly that and mimic `lmer`/`glmer` but that is probably not a sustainable way for doing this. Realistically, I would;t be able to spend time on this for a (long) while.

So, in summary, it would be great but I would need someone to think up a more elegant method for handling non-standard formulas.

**Comment by topepo** [comment]
This is a good idea and is in-process for `tidymodels`. 

Unfortunately, `caret` is at the lifecycle point of having a fairly complete interface, so this isn't a good fit.  

The development work is focused on squashing bugs and any changed required by the underlying model packages.
